# Project description
The commercial department ask to know from the two offered prepaid plans, Surf and Ultimate.
What plan brings in more revenue in order to adjust the advertising budget.
We will carry out a preliminary analysis of the plans based on a relatively small client selection.
The data we have is about 500 Megaline clients. 

## Description of the Megaline plans
Note: Megaline rounds seconds up to minutes, and megabytes to gigabytes. For calls, each individual call is rounded up: even if the call lasted just one second, it will be counted as one minute. For web traffic, individual web sessions are not rounded up. Instead, the total for the month is rounded up. If someone uses 1025 megabytes this month, they will be charged for 2 gigabytes.

#### Surf
- Monthly charge: 20 dollars
- 500 monthly minutes, 50 texts, and 15 GB of data
- After exceeding the package limits:
 - 1 minute: 3 cents
 - 1 text message: 3 cents
 - 1 GB of data: 10 dollars
 
#### Ultimate
- Monthly charge: 70 dollars
- 3000 monthly minutes, 1000 text messages, and 30 GB of data
- After exceeding the package limits:
 - 1 minute: 1 cent
 - 1 text message: 1 cent
 - 1 GB of data: 7 dollars

# Initialization

In [40]:
# Loading all the libraries
import pandas as pd
from pandas_profiling import ProfileReport
import plotly.express as px
import numpy as np

# Load data

In [41]:
# Load the data file into a DataFrame
try:
    calls = pd.read_csv('megaline_calls.csv')
    internet = pd.read_csv('megaline_internet.csv')
    messages = pd.read_csv('megaline_messages.csv')
    plans = pd.read_csv('megaline_plans.csv')
    users = pd.read_csv('megaline_users.csv')
except:
    calls = pd.read_csv('/datasets/megaline_calls.csv')
    internet = pd.read_csv('/datasets/megaline_internet.csv')
    messages = pd.read_csv('/datasets/megaline_messages.csv')
    plans = pd.read_csv('/datasets/megaline_plans.csv')
    users = pd.read_csv('/datasets/megaline_users.csv')

### Explore initial data

#### The datasets contains the following fields:

##### users

The *users* table (data on users):
- user_id — unique user identifier
- first_name — user's name
- last_name — user's last name
- age — user's age (years)
- reg_date — subscription date (dd, mm, yy)
- churn_date — the date the user stopped using the service (if the value is missing, the calling plan -- was being used when this data was retrieved)
- city — user's city of residence
- plan — calling plan name

In [42]:
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   user_id     500 non-null    int64 
 1   first_name  500 non-null    object
 2   last_name   500 non-null    object
 3   age         500 non-null    int64 
 4   city        500 non-null    object
 5   reg_date    500 non-null    object
 6   plan        500 non-null    object
 7   churn_date  34 non-null     object
dtypes: int64(2), object(6)
memory usage: 31.4+ KB


In [43]:
users.head()

user_id first_name  last_name  age                                   city    reg_date      plan churn_date
0     1000   Anamaria      Bauer   45  Atlanta-Sandy Springs-Roswell, GA MSA  2018-12-24  ultimate        NaN
1     1001     Mickey  Wilkerson   28        Seattle-Tacoma-Bellevue, WA MSA  2018-08-13      surf        NaN
2     1002     Carlee    Hoffman   36   Las Vegas-Henderson-Paradise, NV MSA  2018-10-21      surf        NaN
3     1003   Reynaldo    Jenkins   52                          Tulsa, OK MSA  2018-01-28      surf        NaN
4     1004    Leonila   Thompson   40        Seattle-Tacoma-Bellevue, WA MSA  2018-05-23      surf        NaN

##### In this data we have
- Columns number: 8
- Number of entries: 500
- Columns with missing values: 
 - churn_date in this column the values is missing if calling plan was being used when this data was retrieved
- Columns with data type need to be correct: 
 - reg_date to date type

In [44]:
users.describe(include=object)

first_name last_name                                       city    reg_date  plan  churn_date
count         500       500                                        500         500   500          34
unique        458       399                                         73         266     2          29
top       Leonila     David  New York-Newark-Jersey City, NY-NJ-PA MSA  2018-03-08  surf  2018-12-18
freq            3         3                                         80           5   339           3

In [45]:
users.describe()

user_id         age
count   500.000000  500.000000
mean   1249.500000   45.486000
std     144.481833   16.972269
min    1000.000000   18.000000
25%    1124.750000   30.000000
50%    1249.500000   46.000000
75%    1374.250000   61.000000
max    1499.000000   75.000000

In the describe check the relevant thing we learn is:
- the mean age close to the median (45.5 and 46 respectively) the youngest is 18 and the oldest is 75 - reasonable ages.

##### calls

The *calls* table (data on calls):
- id — unique call identifier
- call_date — call date
- duration — call duration (in minutes)
- user_id — the identifier of the user making the call

In [46]:
calls.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 137735 entries, 0 to 137734
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   id         137735 non-null  object 
 1   user_id    137735 non-null  int64  
 2   call_date  137735 non-null  object 
 3   duration   137735 non-null  float64
dtypes: float64(1), int64(1), object(2)
memory usage: 4.2+ MB


In [47]:
calls.head()

id  user_id   call_date  duration
0   1000_93     1000  2018-12-27      8.52
1  1000_145     1000  2018-12-27     13.66
2  1000_247     1000  2018-12-27     14.48
3  1000_309     1000  2018-12-28      5.76
4  1000_380     1000  2018-12-30      4.22

##### In this data we have
- Columns number: 4
- Number of entries: 137735
- Columns with missing values: non 
- Columns with data type need to be correct: 
 - call_date to date type

In [48]:
calls.describe(include=object)

id   call_date
count    137735      137735
unique   137735         351
top     1000_93  2018-12-27
freq          1        1091

In [49]:
calls.describe()

user_id       duration
count  137735.000000  137735.000000
mean     1247.658046       6.745927
std       139.416268       5.839241
min      1000.000000       0.000000
25%      1128.000000       1.290000
50%      1247.000000       5.980000
75%      1365.000000      10.690000
max      1499.000000      37.600000

In the describe check the relevant thing we learn is:
- there are  calls with duration of 0. These might be missed calls. 

##### messages

The *messages* table (data on texts):
- id — unique text message identifier
- message_date — text message date
- user_id — the identifier of the user sending the text

In [50]:
messages.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76051 entries, 0 to 76050
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   id            76051 non-null  object
 1   user_id       76051 non-null  int64 
 2   message_date  76051 non-null  object
dtypes: int64(1), object(2)
memory usage: 1.7+ MB


In [51]:
messages.head()

id  user_id message_date
0  1000_125     1000   2018-12-27
1  1000_160     1000   2018-12-31
2  1000_223     1000   2018-12-31
3  1000_251     1000   2018-12-27
4  1000_255     1000   2018-12-26

##### In this data we have
- Columns number: 3
- Number of entries: 76051
- Columns with missing values: non 
- Columns with data type need to be correct: 
 - message_date to date type

In [52]:
messages.describe(include=object)

id message_date
count      76051        76051
unique     76051          351
top     1000_125   2018-12-28
freq           1          702

In [53]:
messages.describe()

user_id
count  76051.000000
mean    1245.972768
std      139.843635
min     1000.000000
25%     1123.000000
50%     1251.000000
75%     1362.000000
max     1497.000000

In the describe check we have nothing relevant we learn.

##### internet

The *internet* table (data on web sessions):
- id — unique session identifier
- mb_used — the volume of data spent during the session (in megabytes)
- session_date — web session date
- user_id — user identifier

In [54]:
internet.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 104825 entries, 0 to 104824
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   id            104825 non-null  object 
 1   user_id       104825 non-null  int64  
 2   session_date  104825 non-null  object 
 3   mb_used       104825 non-null  float64
dtypes: float64(1), int64(1), object(2)
memory usage: 3.2+ MB


In [55]:
internet.head()

id  user_id session_date  mb_used
0   1000_13     1000   2018-12-29    89.86
1  1000_204     1000   2018-12-31     0.00
2  1000_379     1000   2018-12-28   660.40
3  1000_413     1000   2018-12-26   270.99
4  1000_442     1000   2018-12-27   880.22

In [58]:
internet.describe(include=object)

id session_date
count    104825       104825
unique   104825          351
top     1000_13   2018-12-24
freq          1          851

In [59]:
internet.describe()

user_id        mb_used
count  104825.000000  104825.000000
mean     1242.496361     366.713701
std       142.053913     277.170542
min      1000.000000       0.000000
25%      1122.000000     136.080000
50%      1236.000000     343.980000
75%      1367.000000     554.610000
max      1499.000000    1693.470000

In the describe check the relevant thing we learn is:
 - there are users who didn't use at all the internet plan. We learn this because the min value of mb_used is 0 

##### In this data we have
- Columns number: 4
- Number of entries: 104825
- Columns with missing values: non 
- Columns with data type need to be correct: 
 - session_date to date type

##### plans

The *plans* table (data on the plans):
- plan_name — calling plan name
- usd_monthly_fee — monthly charge in US dollars
- minutes_included — monthly minute allowance
- messages_included — monthly text allowance
- mb_per_month_included — data volume allowance (in megabytes)
- usd_per_minute — price per minute after exceeding the package limits (e.g., if the package includes 100 minutes, the 101st minute will be charged)
- usd_per_message — price per text after exceeding the package limits
- usd_per_gb — price per extra gigabyte of data after exceeding the package limits (1 GB = 1024 megabytes)

In [56]:
plans.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2 entries, 0 to 1
Data columns (total 8 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   messages_included      2 non-null      int64  
 1   mb_per_month_included  2 non-null      int64  
 2   minutes_included       2 non-null      int64  
 3   usd_monthly_pay        2 non-null      int64  
 4   usd_per_gb             2 non-null      int64  
 5   usd_per_message        2 non-null      float64
 6   usd_per_minute         2 non-null      float64
 7   plan_name              2 non-null      object 
dtypes: float64(2), int64(5), object(1)
memory usage: 256.0+ bytes


In [57]:
plans.head()

messages_included  mb_per_month_included  minutes_included  ...  usd_per_message  usd_per_minute  plan_name
0                 50                  15360               500  ...             0.03            0.03       surf
1               1000                  30720              3000  ...             0.01            0.01   ultimate

[2 rows x 8 columns]

##### In this data we have
- Columns number: 8
- Number of entries: 2
- Columns with missing values: non 
- Columns with data type need to be correct: 
 - session_date to date type

No need to run describe test here since we have only 2 rows